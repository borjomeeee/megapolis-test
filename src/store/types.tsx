export const DOWNLOAD_TASKS = "DOWNLOAD_TASKS";
export const DOWNLOAD_TASKS_SUCCESS = "DOWNLOAD_TASKS_SUCCESS";
export const DOWNLOAD_TASKS_FAILED = "DOWNLOAD_TASKS_FAILED";

export const CREATE_TASK = "CREATE_TASK";
export const CREATE_TASK_SUCCESS = "CREATE_TASK_SUCCESS";
export const CREATE_TASK_FAILED = "CREATE_TASK_FAILED";

export const EDIT_TASK = "EDIT_TASK";
export const EDIT_TASK_SUCCESS = "EDIT_TASK_SUCCESS";
export const EDIT_TASK_FAILED = "EDIT_TASK_FAILED";

export const REMOVE_TASK = "REMOVE_TASK";
export const REMOVE_TASK_SUCCESS = "REMOVE_TASK_SUCCESS";
export const REMOVE_TASK_FAILED = "REMOVE_TASK_FAILED";
