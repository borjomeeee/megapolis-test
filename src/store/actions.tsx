import * as TYPES from "./types";

import { ITodoItem } from "../models/TodoItem.model";

// Download
export const downloadTasksAction = () =>
  ({
    type: TYPES.DOWNLOAD_TASKS,
  } as const);

export const downloadTasksSuccessAction = (tasks: ITodoItem[]) =>
  ({
    type: TYPES.DOWNLOAD_TASKS_SUCCESS,
    payload: { tasks },
  } as const);

export const downloadTasksFailedAction = (errMsg: string) =>
  ({
    type: TYPES.DOWNLOAD_TASKS_FAILED,
    payload: { errMsg },
  } as const);

// Create
export const createTaskAction = (descr: string) =>
  ({
    type: TYPES.CREATE_TASK,
    payload: { descr },
  } as const);

export const createTaskSuccessAction = (task: ITodoItem) =>
  ({
    type: TYPES.CREATE_TASK_SUCCESS,
    payload: { task },
  } as const);

export const createTaskFailedAction = (errMsg: string) =>
  ({
    type: TYPES.CREATE_TASK_FAILED,
    payload: { errMsg },
  } as const);

// Edit
export const editTaskAction = (id: number, descr: string) =>
  ({
    type: TYPES.EDIT_TASK,
    payload: { id, descr },
  } as const);

export const editTaskSuccessAction = (id: number, descr: string) =>
  ({
    type: TYPES.EDIT_TASK_SUCCESS,
    payload: { id, descr },
  } as const);

export const editTaskFailedAction = (errMsg: string) =>
  ({
    type: TYPES.EDIT_TASK_FAILED,
    payload: { errMsg },
  } as const);

// Remove
export const removeTaskAction = (id: number) =>
  ({
    type: TYPES.REMOVE_TASK,
    payload: { id },
  } as const);

export const removeTaskSuccessAction = (id: number) =>
  ({
    type: TYPES.REMOVE_TASK_SUCCESS,
    payload: { id },
  } as const);

export const removeTaskFailedAction = (errMsg: string) =>
  ({
    type: TYPES.REMOVE_TASK_FAILED,
    payload: { errMsg },
  } as const);
