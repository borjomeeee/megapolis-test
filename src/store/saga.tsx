import { takeLeading, put } from "redux-saga/effects";

import { DOWNLOAD_TASKS, CREATE_TASK, EDIT_TASK, REMOVE_TASK } from "./types";
import {
  downloadTasksSuccessAction,
  createTaskSuccessAction,
  editTaskSuccessAction,
  removeTaskSuccessAction,
  createTaskAction,
  editTaskAction,
  removeTaskAction,
} from "./actions";

function* downloadTasksSaga() {
  yield put(
    downloadTasksSuccessAction([
      { id: 1, descr: "Описание 1" },
      { id: 2, descr: "Описание 2" },
    ])
  );
}

function* createTaskSaga({ payload }: ReturnType<typeof createTaskAction>) {
  yield put(createTaskSuccessAction({ id: 3, descr: payload.descr }));
}

function* editTaskSaga({ payload }: ReturnType<typeof editTaskAction>) {
  yield put(editTaskSuccessAction(payload.id, payload.descr));
}

function* removeTaskSaga({ payload }: ReturnType<typeof removeTaskAction>) {
  yield put(removeTaskSuccessAction(payload.id));
}

export default function* rootSaga() {
  yield takeLeading(DOWNLOAD_TASKS, downloadTasksSaga);
  yield takeLeading(CREATE_TASK, createTaskSaga);
  yield takeLeading(EDIT_TASK, editTaskSaga);
  yield takeLeading(REMOVE_TASK, removeTaskSaga);
}
