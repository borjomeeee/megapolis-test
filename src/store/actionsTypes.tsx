import * as ACTIONS from "./actions";

export type IAction =
  | ReturnType<typeof ACTIONS.downloadTasksAction>
  | ReturnType<typeof ACTIONS.downloadTasksSuccessAction>
  | ReturnType<typeof ACTIONS.downloadTasksFailedAction>
  | ReturnType<typeof ACTIONS.createTaskAction>
  | ReturnType<typeof ACTIONS.createTaskSuccessAction>
  | ReturnType<typeof ACTIONS.createTaskFailedAction>
  | ReturnType<typeof ACTIONS.editTaskAction>
  | ReturnType<typeof ACTIONS.editTaskSuccessAction>
  | ReturnType<typeof ACTIONS.editTaskFailedAction>
  | ReturnType<typeof ACTIONS.removeTaskAction>
  | ReturnType<typeof ACTIONS.removeTaskSuccessAction>
  | ReturnType<typeof ACTIONS.removeTaskFailedAction>;
