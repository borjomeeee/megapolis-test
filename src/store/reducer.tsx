import { initialState, IInitialState } from ".";

import { IAction } from "./actionsTypes";

export default (
  state: IInitialState = initialState,
  action: IAction
): IInitialState => {
  switch (action.type) {
    default:
      return state;
  }
};
