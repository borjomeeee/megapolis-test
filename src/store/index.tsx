import { createStore, applyMiddleware } from "redux";
import { composeWithDevTools } from "redux-devtools-extension";
import createSagaMiddleware from "redux-saga";

import rootSaga from "./saga";

import reducer from "./reducer";

import { ITodoItem } from "../models/TodoItem.model";

export const initialState = {
  tasks: [
    { id: 1, descr: "Описание 1" },
    { id: 2, descr: "Описание 2" },
    { id: 3, descr: "Описание 3" },
  ] as ITodoItem[],
  error: "",
  isLoading: true,
};
export type IInitialState = typeof initialState;

const sagaMiddleware = createSagaMiddleware();

const store = createStore(
  reducer,
  composeWithDevTools(applyMiddleware(sagaMiddleware))
);
sagaMiddleware.run(rootSaga);

export default store;
