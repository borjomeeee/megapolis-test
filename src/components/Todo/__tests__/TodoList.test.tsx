import React from "react";
import { Store, createStore } from "redux";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";

import { shallow, mount } from "enzyme";

import renderer from "react-test-renderer";

import reducer, { IInitialState } from "../../../store/reducer";

import TodoListComponent from "../TodoList.component";

import { IAction } from "../../../store/actionsTypes";

describe("todo list tests", () => {
  it("render correctly", () => {
    const initialState: IInitialState = {
      tasks: [
        { id: 1, descr: "Descr 1" },
        { id: 2, descr: "Descr 2" },
        { id: 3, descr: "Descr 3" },
      ],
      error: "",
      isLoading: false,
    };

    const mockStore = configureStore();
    let store = mockStore(initialState);

    const component = renderer.create(
      <Provider store={store}>
        <TodoListComponent />
      </Provider>
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe("test reducer actions", () => {
    const store = createStore(reducer);

    store.dispatch({
      type: "DOWNLOAD_TASKS_SUCCESS",
      payload: {
        tasks: [
          { id: 1, descr: "Descr 1" },
          { id: 2, descr: "Descr 2" },
          { id: 3, descr: "Descr 3" },
        ],
      },
    });

    it("correct add task", () => {
      let wrapper = mount(
        <Provider store={store}>
          <TodoListComponent />
        </Provider>
      );

      expect(wrapper.find("TodoListItemComponent")).toHaveLength(3);

      store.dispatch({
        type: "CREATE_TASK_SUCCESS",
        payload: { task: { id: 4, descr: "Descr 4" } },
      });

      wrapper = mount(
        <Provider store={store}>
          <TodoListComponent />
        </Provider>
      );

      expect(wrapper.find("TodoListItemComponent")).toHaveLength(4);
    });

    it("correct remove task", () => {
      let wrapper = mount(
        <Provider store={store}>
          <TodoListComponent />
        </Provider>
      );

      expect(wrapper.find("TodoListItemComponent")).toHaveLength(4);

      store.dispatch({
        type: "REMOVE_TASK_SUCCESS",
        payload: { id: 3 },
      });

      wrapper = mount(
        <Provider store={store}>
          <TodoListComponent />
        </Provider>
      );

      expect(wrapper.find("TodoListItemComponent")).toHaveLength(3);
    });
  });

  it("test empty tasks list", () => {
    const store = createStore(reducer);

    let wrapper = mount(
      <Provider store={store}>
        <TodoListComponent />
      </Provider>
    );

    expect(wrapper.find("TodoListItemComponent")).toHaveLength(0);
    expect(wrapper.find("div").text()).toBe("Пустой лист тасков!");
  });
});
