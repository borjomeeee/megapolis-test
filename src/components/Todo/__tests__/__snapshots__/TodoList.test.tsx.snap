// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`todo list tests render correctly 1`] = `
<table
  className="todo-list"
>
  <tbody>
    <tr
      className="todo-item todo-list__row"
    >
      <td
        className="todo-item__col"
      >
        Задача №1
      </td>
      <td
        className="todo-item__col"
        width="70%"
      >
        Descr 1
      </td>
      <td
        className="todo-item__col"
        width="100px"
      >
        <div
          className="todo-item__change-container"
        >
          <div
            className="todo-item__icon todo-item__icon_edit"
          >
            <IconSvg
              onClick={[Function]}
            />
          </div>
          <div
            className="todo-item__icon todo-item__icon_remove"
          >
            <IconSvg
              onClick={[Function]}
            />
          </div>
        </div>
      </td>
    </tr>
    <tr
      className="todo-item todo-list__row"
    >
      <td
        className="todo-item__col"
      >
        Задача №2
      </td>
      <td
        className="todo-item__col"
        width="70%"
      >
        Descr 2
      </td>
      <td
        className="todo-item__col"
        width="100px"
      >
        <div
          className="todo-item__change-container"
        >
          <div
            className="todo-item__icon todo-item__icon_edit"
          >
            <IconSvg
              onClick={[Function]}
            />
          </div>
          <div
            className="todo-item__icon todo-item__icon_remove"
          >
            <IconSvg
              onClick={[Function]}
            />
          </div>
        </div>
      </td>
    </tr>
    <tr
      className="todo-item todo-list__row"
    >
      <td
        className="todo-item__col"
      >
        Задача №3
      </td>
      <td
        className="todo-item__col"
        width="70%"
      >
        Descr 3
      </td>
      <td
        className="todo-item__col"
        width="100px"
      >
        <div
          className="todo-item__change-container"
        >
          <div
            className="todo-item__icon todo-item__icon_edit"
          >
            <IconSvg
              onClick={[Function]}
            />
          </div>
          <div
            className="todo-item__icon todo-item__icon_remove"
          >
            <IconSvg
              onClick={[Function]}
            />
          </div>
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;
